src/dstore/pmix_esh.c

```diff
@@ -0,0 +1,1374 @@
+            data_segment_size = (size_t)page_size;
+        }
+    }
+    if (0 == data_segment_size) {
+        data_segment_size = NS_DATA_SEG_SIZE;
+    }
+    if (NULL != (str = getenv("SM_USE_LINEAR_SEARCH"))) {
+        if (1 == strtoul(str, NULL, 10)) {
+            direct_mode = 1;
+        }
+    }
+}
+
+static void _delete_sm_desc(seg_desc_t *desc)
+{
+    seg_desc_t *tmp;
+    /* free all global segments */
+    while (NULL != desc) {
+        tmp = desc->next;
+        /* detach & unlink from current desc */
+        if (desc->seg_info.seg_cpid == getpid()) {
+            pmix_sm_segment_unlink(&desc->seg_info);
+        }
+        pmix_sm_segment_detach(&desc->seg_info);
+        free(desc);
+        desc = tmp;
+    }
+}
+
+static int _pmix_getpagesize(void)
+{
+#if defined(_SC_PAGESIZE )
+    return sysconf(_SC_PAGESIZE);
+#elif defined(_SC_PAGE_SIZE)
+    return sysconf(_SC_PAGE_SIZE);
+#else
+    return 65536; /* safer to overestimate than under */
+#endif
+}
+
+static seg_desc_t *_create_new_segment(segment_type type, char *nsname, uint32_t id)
+{
+    int rc;
+    char file_name[PMIX_PATH_MAX];
+    size_t size;
+    seg_desc_t *new_seg = NULL;
+
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s: segment type %d, nspace %s, id %u", __FILE__, __LINE__, __func__, type, nsname, id));
+
+    switch (type) {
+        case INITIAL_SEGMENT:
+            size = initial_segment_size;
+            snprintf(file_name, PMIX_PATH_MAX, "%s/initial-pmix_shared-segment-%u", tmp_dir, id);
+            break;
+        case NS_META_SEGMENT:
+            size = meta_segment_size;
+            snprintf(file_name, PMIX_PATH_MAX, "%s/smseg-%s-%u", tmp_dir, nsname, id);
+            break;
+        case NS_DATA_SEGMENT:
+            size = data_segment_size;
+            snprintf(file_name, PMIX_PATH_MAX, "%s/smdataseg-%s-%d", tmp_dir, nsname, id);
+            break;
+        default:
+            PMIX_ERROR_LOG(PMIX_ERROR);
+            return NULL;
+    }
+    new_seg = (seg_desc_t*)malloc(sizeof(seg_desc_t));
+    if (new_seg) {
+        new_seg->id = id;
+        new_seg->next = NULL;
+        new_seg->type = type;
+        rc = pmix_sm_segment_create(&new_seg->seg_info, file_name, size);
+        if (PMIX_SUCCESS == rc) {
+            memset(new_seg->seg_info.seg_base_addr, 0, size);
+        } else {
+            free(new_seg);
+            new_seg = NULL;
+            PMIX_ERROR_LOG(rc);
+        }
+    }
+    return new_seg;
+}
+
+static seg_desc_t *_attach_new_segment(segment_type type, char *nsname, uint32_t id)
+{
+    int rc;
+    seg_desc_t *new_seg = NULL;
+    new_seg = (seg_desc_t*)malloc(sizeof(seg_desc_t));
+    new_seg->id = id;
+    new_seg->next = NULL;
+    new_seg->type = type;
+
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s: segment type %d, nspace %s, id %u", __FILE__, __LINE__, __func__, type, nsname, id));
+
+    switch (type) {
```

Comment:
If my understanding is correct we have the same `switch` logic in `_attach_new_segment()` and `_create_new_segment()`.
I suggest to put this switch to a separate function so we'll have a code reuse.